lib/TPP/ToBlockLayoutAndBack.cpp

```diff
@@ -776,9 +776,169 @@ struct PropagateThroughElementWiseOp
+        blockingFactors(blockingFactors) {}
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    FailureOr<linalg::GenericOp> packedMatmul = mlir::linalgx::packMatmulOp(
+        rewriter, matmulOp,
+        getAsOpFoldResult(rewriter.getI64ArrayAttr(blockingFactors)));
+    if (failed(packedMatmul))
+      return failure();
+    return success();
+  }
+
+private:
+  ArrayRef<int64_t> blockingFactors;
+};
+
+// From linalg.generic to linalg.matmul.
+struct DeGeneralizeMatmul : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!linalgOp.hasTensorSemantics())
+      return failure();
+    if (!tpp::utils::isMarkedWithTpp(linalgOp, "tpp.matmul"))
+      return failure();
+    SmallVector<Value> inputOperands = linalgOp.getDpsInputOperands();
+    SmallVector<Value> outputOperands = linalgOp.getDpsInitOperands();
+    rewriter.replaceOpWithNewOp<linalg::MatmulOp>(
+        linalgOp, linalgOp.getResultTypes(), inputOperands, outputOperands);
+    return success();
+  }
+};
+
+// Entry point for packing a matmul operation.
+// Pack MatmulOp as following:
+// [NB][KB][nb][kb] += [NB][CB][nb][cb] * [KB][CB][cb][kb]
+// CB = batch reduce dimension.
+struct PackMatmul : public PackMatmulBase<PackMatmul> {
+  PackMatmul() = default;
+  PackMatmul(ArrayRef<int64_t> blockingFactors) {
+    this->blockingFactors = blockingFactors;
+  }
+
+  void runOnOperation() override {
+    if (blockingFactors.empty())
+      return;
+    MLIRContext *ctx = getOperation().getContext();
+    RewritePatternSet patterns(ctx);
+    mlir::tpp::populateSinkPackPatterns(patterns);
+    patterns.add<DoItOnMatmul>(ctx, blockingFactors);
+    patterns.add<DeGeneralizeMatmul>(ctx);
+    (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
+    return;
+  }
+};
+
+struct DoItOnConv2DNchwFchw
+    : public OpRewritePattern<linalg::Conv2DNchwFchwOp> {
+  DoItOnConv2DNchwFchw(MLIRContext *context, ArrayRef<int64_t> blockingFactors,
+                       PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::Conv2DNchwFchwOp>(context, benefit),
+        blockingFactors(blockingFactors) {}
+
+  LogicalResult matchAndRewrite(linalg::Conv2DNchwFchwOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    FailureOr<linalg::GenericOp> genericOp =
+        mlir::linalgx::packConv2DNchwFchwOp(
+            rewriter, linalgOp,
+            getAsOpFoldResult(rewriter.getI64ArrayAttr(blockingFactors)));
+    if (failed(genericOp))
+      return failure();
+    return success();
+  }
+
+private:
+  SmallVector<int64_t> blockingFactors;
+};
+
+struct PackConv2DNchwFchw : public PackConv2DNchwFchwBase<PackConv2DNchwFchw> {
```

Comment:
What about the tests for these transforms? Are they all using transforms now? If so, why are we bringing this code back if it's not going to be tested?

Reply:
Do you want to use transform for packing?